core/src/main/java/org/elasticsearch/search/aggregations/pipeline/movavg/SimulatedAnealingMinimizer.java

```diff
@@ -0,0 +1,129 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.pipeline.movavg;
+
+import com.google.common.collect.EvictingQueue;
+import org.elasticsearch.search.aggregations.pipeline.movavg.models.MovAvgModel;
+
+/**
+ * A cost minimizer which will fit a MovAvgModel to the data.
+ *
+ * This optimizer uses naive simulated annealing.  Random solutions in the problem space
+ * are generated, compared against the last period of data, and the least absolute deviation
+ * is recorded as a cost.
+ *
+ * If the new cost is better than the old cost, the new coefficients are chosen.  If the new
+ * solution is worse, there is a temperature-dependent probability it will be randomly selected
+ * anyway.  This allows the algo to sample the problem space widely.  As iterations progress,
+ * the temperature decreases and the algorithm rejects poor solutions more regularly,
+ * theoretically honing in on a global minimum.
+ */
+public class SimulatedAnealingMinimizer {
+
+    /**
+     * Runs the simulated annealing algorithm and produces a new set of model coefficients.
+     * The method is quasi-generic, accepting any number of coefficients and trusting that
+     * the underlying model knows what to do with them
+     *
+     * @param model         The MovAvgModel to be optimized for
+     * @param train         A training set provided to the model, which predictions will be
+     *                      generated from
+     * @param test          A test set of data to compare the predictions against and derive
+     *                      a cost per set of coefficeints
+     * @return              The new set of model coefficients
+     */
+    public static MovAvgModel minimize(MovAvgModel model, EvictingQueue<Double> train, double[] test) {
+
+        double temp = 1;
+        double minTemp = 0.0001;
+        int iterations = 100;
+        double alpha = 0.9;
+
+        MovAvgModel bestModel = model;
+        MovAvgModel oldModel = model;
+        MovAvgModel newModel;
+
+        double oldCost = cost(model, train, test);
+        double bestCost = oldCost;
+        double newCost;
+        double ap;
+
+        while (temp > minTemp) {
+            for (int i = 0; i < iterations; i++) {
+                newModel = oldModel.neighboringModel();
+                newCost = cost(newModel, train, test);
+
+                ap = acceptanceProbability(oldCost, newCost, temp);
+                if (ap > Math.random()) {
+                    oldModel = newModel.clone();
+                    oldCost = newCost;
+
+                    if (newCost < bestCost) {
+                        bestCost = newCost;
+                        bestModel = newModel.clone();
```

Comment:
I'm not sure the `clone()` for `bestModel` and `oldModel` are required here as `neighbouringModel()` creates a new instance of the model anyway so there is no way that the parameters in `bestModel` and `oldModel` could be changed?